### PR TITLE
feat(openapi): support versioned docs

### PR DESCRIPTION
**Pull request by zchsh**
## 🔗 Relevant links

- [Preview link][preview] 🔎
- [Asana task][task] 🎟️

## 🗒️ What

Implements support for versioned `OpenApiDocsView` pages.

## 🤷 Why

To allow visitors to `OpenAPIDocsView` pages to browse past versions of API documentation. Some users may need to reference an older version of the API, and it's important we support them.

## 🛠️ How

- fix: width style issue with version switcher
- feat: add versionSwitcherSlot in OpenApiOverview
- feat: implement get-version-switcher-props
- feat: generate versionSwitcherProps, pass to view

## 📸 Design Screenshots

No visual changes to existing pages.

## 🧪 Testing

- [ ] Visit [/hcp/api-docs/vault-secrets], ensure the page renders as it does upstream
    - HCP Vault Secrets does not yet have multiple versions of its spec, so this new feature should have no effect on the page
- [ ] Test out a set of versioned OpenAPI docs locally
    - See [this Slack message](https://hashicorp.slack.com/archives/C02T80JTNS2/p1695668653082409) for instructions on how to test this

## 💭 Anything else?

Not at the moment!

[task]: https://app.asana.com/0/1204678746647847/1205062681720537/f
[preview]: https://dev-portal-git-zsopenapi-versioned-docs-hashicorp.vercel.app/
[/hcp/api-docs/vault-secrets]: https://dev-portal-git-zsopenapi-versioned-docs-hashicorp.vercel.app/hcp/api-docs/vault-secrets

### CODE REVIEW

File: src/components/version-switcher/version-switcher.module.css [hunk 1]
```diff
@@ -28,7 +28,8 @@ https://app.asana.com/0/1202097197789424/1204412156894162/f
 
 	/* Make dropdown list fill container */
 	& > div {
-		width: 100%;
+		min-width: 100%;
+		width: max-content;
```
Comment (zchsh) [inline src/components/version-switcher/version-switcher.module.css:32]: This fixes an issue where the `VersionSwitcher` [listPosition](https://github.com/hashicorp/dev-portal/blob/zs.openapi-versioned-docs/src/components/version-switcher/index.tsx#L39) being passed to our dropdown disclosure component wasn't working as expected (due to interactions with these "overrides).

Note this change fixes _all_ uses of version switcher, not just the use in `OpenApiDocsView`.

| Before | After |
| - | - |
| version-switcher-before | version-switcher-after |

File: src/views/open-api-docs-view/components/open-api-overview/index.tsx [hunk 2]
```diff
@@ -42,31 +43,37 @@ export function OpenApiOverview({
...
+					</span>
+					<Badge
+						className={s.releaseStageBadge}
+						text={badgeText}
+						type="outlined"
+						size="small"
+					/>
+				</header>
+				{versionSwitcherSlot ? (
+					<div className={s.versionSwitcherSlot}>{versionSwitcherSlot}</div>
```
Comment (zchsh) [inline src/views/open-api-docs-view/components/open-api-overview/index.tsx:74]: The `OpenApiOverview` component now supports an optional `versionSwitcherSlot`. The work in this component is refactoring the layout a bit to support this change. I've used a `@container` query to try to keep the styling easy to understand across viewports.

File: src/views/open-api-docs-view/index.tsx [hunk 3]
```diff
@@ -62,6 +64,14 @@ function OpenApiDocsView({
 						badgeText={releaseStage}
 						serviceProductSlug={serviceProductSlug}
 						statusIndicatorConfig={statusIndicatorConfig}
+						versionSwitcherSlot={
+							versionSwitcherProps ? (
```
Comment (zchsh) [inline src/views/open-api-docs-view/index.tsx:68]: At the view level, it's a relatively simple task of render a `VersionSwitcher` with `versionSwitcherProps` into the overview component (if those props are present, which they'll only be for OpenAPI docs with 2 or more separate versions).

File: src/views/open-api-docs-view/utils/get-version-switcher-props.ts [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+import type { VersionSwitcherProps } from 'components/version-switcher'
+import type { OpenApiDocsVersionData } from '../types'
+
+/**
+ * Given version data and other OpenAPI docs details,
+ * Return version switcher dropdown props for use in `OpenApiDocsView`.
+ *
+ * Note: If there is only one version, we return null.
+ */
+export function getVersionSwitcherProps({
```
Comment (zchsh) [inline src/views/open-api-docs-view/utils/get-version-switcher-props.ts:16]: This function does the server-side work of transforming the fetched version data (as well as some basic OpenAPI spec info for things like the project name) into a set of `VersionSwitcherProps`.

Note this work builds on versioned API docs implemented for HCP Packer. Namely, [fetch-cloud-api-version-data](https://github.com/hashicorp/dev-portal/blob/main/src/lib/api-docs/fetch-cloud-api-version-data/index.ts). The fetching of this version data happens at the _page_ level (this allows us to take different approaches to fetching versioned OpenAPI docs data if we want, as this is expected to differ between HCP APIs and other API docs!).

File: src/views/open-api-docs-view/components/open-api-overview/open-api-overview.module.css [hunk 5]
```diff
@@ -7,6 +7,22 @@
...
+	container-type: inline-size;
+}
+
+.headerAndVersionSwitcher {
+	display: flex;
+	flex-direction: column-reverse;
+	gap: 24px;
+
+	/* stylelint-disable-next-line at-rule-no-unknown */
+	@container (min-width: 640px) {
```
Comment (kendallstrautman) [inline src/views/open-api-docs-view/components/open-api-overview/open-api-overview.module.css:21]: Fun! I've never used one of these 💖

**Comment by kendallstrautman** [comment]
Screenshot 2023-09-26 at 3 51 47 PM
Do we feel the need to add a banner or indication of when youre viewing a version that isn't latest. I believe we do this on docs? Not a blocker per say, just thinking through the UX

File: src/components/version-switcher/version-switcher.module.css (hunk 1 above)
Comment (kendallstrautman) [reply src/components/version-switcher/version-switcher.module.css:32]: Oh awesome! nice find.

File: src/views/open-api-docs-view/components/open-api-overview/index.tsx (hunk 2 above)
Comment (kendallstrautman) [reply src/views/open-api-docs-view/components/open-api-overview/index.tsx:74]: Slots for the win 🦾

**Review by kendallstrautman — APPROVED** [review]
Testing things locally based on the slack message and works great! 💯 Makes sense to me and I like how simple this is just based on the filesystem organization.

**Comment by zchsh** [comment]
@kendallstrautman Ah great call re: the alert banner! I've added a follow-up task here, we have [prior art via ApiDocsVersionAlert](https://github.com/hashicorp/dev-portal/blob/e77bcd7173963a7430221f92fa2d5559c275848a/src/views/api-docs-view/index.tsx#L55) so I think this should be pretty feasible (more details in task!):

- 🎟 [Add ApiDocsVersionAlert banner to versioned OpenApiDocsView pages](https://app.asana.com/0/1204678746647847/1205587495423439/f)

Incidentally, this reminded me I needed to add `NoIndexTagIfVersioned` here for versioned pages! Just pushed a commit now that does that. Will merge this in after rebasing on [dev-portal#2182 - chore(openapi): fix up duplicate VersionData types](https://github.com/hashicorp/dev-portal/pull/2182) 👍

**Pull request merged**